packages/server/src/koaJsonRpc/index.ts

```diff
@@ -119,80 +134,162 @@ export default class KoaJsonRpc {
+          break;
 
-      const methodTotalLimit = this.registryTotal[methodName];
-      if (this.rateLimit.shouldRateLimit(ctx.ip, methodName, methodTotalLimit, this.requestId)) {
-        ctx.body = jsonResp(body.id, new IPRateLimitExceeded(methodName), undefined);
-        ctx.status = 409;
-        ctx.state.status = `${ctx.status} (${IP_RATE_LIMIT_EXCEEDED})`;
-        return;
-      }
+        // CONTRACT_REVERT
+        case -32008:
+          ctx.status = 200;
+          ctx.state.status = `${ctx.status} (${CONTRACT_REVERT})`;
+          break;
 
-      try {
-        result = await this.registry[body.method](body.params);
-        ctx.state.status = responseSuccessStatusCode;
-      } catch (e: any) {
-        if (e instanceof InvalidParamsError) {
-          ctx.body = jsonResp(body.id, new InvalidParamsError(e.message), undefined);
+        // INVALID_REQUEST
+        case -32600:
+          ctx.status = 400;
+          ctx.state.status = `${ctx.status} (${INVALID_REQUEST})`;
+          break;
+
+        // INVALID_PARAMS
+        case -32602:
           ctx.status = 400;
           ctx.state.status = `${ctx.status} (${INVALID_PARAMS_ERROR})`;
-          return;
-        }
-        ctx.body = jsonResp(body.id, new InternalError(e.message), undefined);
-        ctx.status = 500;
-        ctx.state.status = `${ctx.status} (${INTERNAL_ERROR})`;
-        return;
+          break;
+
+        // METHOD_NOT_FOUND
+        case -32601:
+          ctx.status = 400;
+          ctx.state.status = `${ctx.status} (${METHOD_NOT_FOUND})`;
+          break;
+
+        // IP_RATE_LIMIT_EXCEEDED
+        case -32605:
+          ctx.status = 409;
+          ctx.state.status = `${ctx.status} (${IP_RATE_LIMIT_EXCEEDED})`;
+          break;
+
+        // ANYTHING ELSE
+        default:
+          ctx.status = 400;
+          ctx.state.status = `${ctx.status} (${JSON_RPC_ERROR})`;
+          break;
       }
+    }
+  }
+
+  private async handleMultipleRequest(ctx, body: any): Promise<void> {
+    // verify that batch requests are enabled
+    if (!this.getBatchRequestsEnabled()) {
+      ctx.body = jsonResp(null, predefined.BATCH_REQUESTS_DISABLED, undefined);
+      ctx.status = 400;
+      ctx.state.status = `${ctx.status} (${INVALID_REQUEST})`;
+      return;
+    }
+
+    // verify max batch size
+    if (body.length > this.batchRequestsMaxSize) {
+      ctx.body = jsonResp(
+        null,
+        predefined.BATCH_REQUESTS_AMOUNT_MAX_EXCEEDED(body.length, this.batchRequestsMaxSize),
+        undefined,
+      );
+      ctx.status = 400;
+      ctx.state.status = `${ctx.status} (${INVALID_REQUEST})`;
+      return;
+    }
+
+    const response: any[] = [];
+    ctx.state.methodName = 'batch_requests';
+
+    // we do the requests in parallel to save time, but we need to keep track of the order of the responses (since the id might be optional)
+    const promises = body.map((item: any) => this.getRequestResult(item, ctx.ip));
```

Comment:
Q: should `handleMultipleRequest()` be calling `handleSingleRequest()` so simplify and optimize code share?
If it's a large refactor let's open a ticket to follow up

Reply:
It can't I placed all the shared code on `getRequestResult` and basically the difference between `handleMultipleRequest` and `handleSingleRequest` is that the first executes a batch of calls to `getRequestResult` and then response `200` http status with the collection of responses, as opposed `handleSingleRequest` keeps current behavior intact and depending on the result of the request changes the `http status` code response accordingly to the switch that has after the `getRequestResult`. hope this makes sense.